### PR TITLE
feat: reth's block body fns

### PR DESCRIPTION
**Pull request by stevencartavia**
## Motivation
https://github.com/paradigmxyz/reth/issues/12519

## Solution
Brought BlockBody fns from reth to replace reth's BlockBody type with alloy's

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Comment by stevencartavia** [comment]
I just brought these fns at the moment since the rest explicitly use `T: TransactionSigned`

File: crates/consensus/src/transaction/mod.rs [hunk 1]
```diff
@@ -177,6 +180,18 @@ pub trait Transaction: fmt::Debug + any::Any + Send + Sync + 'static {
...
+    #[inline]
+    fn is_eip4844(&self) -> bool {
+        self.ty() == EIP4844_TX_TYPE_ID
+    }
+
+    /// Returns true if the transaction is an EIP-7702 transaction.
+    #[inline]
+    fn is_eip7702(&self) -> bool {
+        self.ty() == EIP7702_TX_TYPE_ID
+    }
```
Comment (mattsse) [inline crates/consensus/src/transaction/mod.rs:194]: these are no longer necessary with #1773

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, makes sense

one suggestion

File: crates/consensus/src/block/mod.rs [hunk 2]
```diff
@@ -58,6 +60,56 @@ impl<T> BlockBody<T> {
...
+    #[inline]
+    pub fn has_eip7702_transactions(&self) -> bool {
+        self.transactions.iter().any(|tx| tx.is_eip7702())
+    }
+
+    /// Helper for blob_transactions_iter function
+    #[inline]
+    fn is_eip4844(tx: &T) -> bool {
+        tx.is_eip4844()
+    }
```
Comment (mattsse) [inline crates/consensus/src/block/mod.rs:100]: this function is not needed and can be removed

File: crates/consensus/src/block/mod.rs [hunk 3]
```diff
@@ -58,6 +60,56 @@ impl<T> BlockBody<T> {
...
+    #[inline]
+    pub fn blob_transactions_iter(&self) -> impl Iterator<Item = &T> + '_ {
+        self.transactions.iter().filter(|tx| Self::is_eip4844(tx))
+    }
+
+    /// Returns only the blob transactions, if any, from the block body.
+    #[inline]
+    pub fn blob_transactions(&self) -> Vec<&T> {
+        self.blob_transactions_iter().collect()
+    }
```
Comment (mattsse) [inline crates/consensus/src/block/mod.rs:112]: this should be an impl Iterator instead because this is mostly useful when iterating over blob txs

**Review by mattsse — CHANGES_REQUESTED** [review]
nits

**Review by mattsse — APPROVED** [review]

**Pull request merged**